File: src/hxcore/flecs/flecs.ts
import { dlopen, ptr } from "bun:ffi";

import { fileURLToPath } from "url";
import { dirname } from "path";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

const libPath = `${__dirname}/bin/libflecswrapper.so`;
console.log("Loading libflecswrapper from", libPath);

const lib = dlopen(libPath, {
  // Lifecycle
  flecs_init: { args: [], returns: "void" },
  flecs_progress: { args: ["float"], returns: "void" },
  flecs_fini: { args: [], returns: "void" },
  flecs_version: { args: [], returns: "cstring" },

  // Component management
  flecs_component_get_id_by_name: { args: ["cstring"], returns: "int32_t" },
  flecs_component_is_mark_changed_by_name: { args: ["uint32_t", "cstring"], returns: "bool" },
  flecs_component_is_marked_changed: { args: ["uint32_t", "uint32_t"], returns: "bool" },
  flecs_component_is_tag: { args: ["uint32_t"], returns: "bool" },
  flecs_component_print_registry: { args: [], returns: "void" },

  // Entity component inspection
  flecs_entity_print_components: { args: ["uint32_t"], returns: "void" },
  flecs_entity_has_component: { args: ["uint32_t", "uint32_t"], returns: "bool" },
  flecs_entity_has_component_by_name: { args: ["uint32_t", "cstring"], returns: "bool" },

  // Entity component modification
  flecs_entity_add_component: { args: ["uint32_t", "uint32_t"], returns: "bool" },
  flecs_entity_add_component_by_name: { args: ["uint32_t", "cstring"], returns: "bool" },
  flecs_entity_remove_component: { args: ["uint32_t", "uint32_t"], returns: "bool" },
  flecs_entity_remove_component_by_name: { args: ["uint32_t", "cstring"], returns: "bool" },

  // Specific component helpers
  flecs_entity_set_velocity: { args: ["uint32_t", "float", "float"], returns: "bool" },
  flecs_entity_get_velocity: { args: ["uint32_t", "pointer", "pointer"], returns: "bool" },
  flecs_entity_set_position: { args: ["uint32_t", "float", "float"], returns: "bool" },
  flecs_entity_get_position: { args: ["uint32_t", "pointer", "pointer"], returns: "bool" },
  flecs_entity_set_destination: { args: ["uint32_t", "float", "float", "float"], returns: "bool" },
  flecs_entity_get_destination: { args: ["uint32_t", "pointer", "pointer", "pointer"], returns: "bool" },

  // Generic vec2 component helpers
  flecs_entity_set_component_vec2: { args: ["uint32_t", "uint32_t", "float", "float"], returns: "bool" },
  flecs_entity_get_component_vec2: { args: ["uint32_t", "uint32_t", "pointer", "pointer"], returns: "bool" },

  // Entity lifecycle
  flecs_entity_create: { args: ["cstring"], returns: "uint32_t" },
  flecs_entity_destroy: { args: ["uint32_t"], returns: "bool" },

  // Observer registration (callbacks not supported directly in Bun FFI)
  // flecs_register_observer: { ... },
  // System registration (callbacks not supported directly in Bun FFI)
  // flecs_register_system: { ... },
});

console.log("Loaded libflecswrapper");

function stringToPtr(str: string) {
  const buf = Buffer.from(str + "\0", "utf8");
  return ptr(buf);
}

const x_buf = new Float32Array(1);
const y_buf = new Float32Array(1);
const z_buf = new Float32Array(1);

function get_vec2(entity: number, fn: (eid: number, xp: any, yp: any) => boolean) {
  return fn(entity, x_buf, y_buf) ? { x: x_buf[0], y: y_buf[0] } : null;
}

function get_entity_component_vec2_by_name(entity: number, component: string) {
  const id = lib.symbols.flecs_get_component_id_by_name(stringToPtr(component));
  if (!id) return null;

  const success = lib.symbols.flecs_get_entity_component_vec2(entity, id, x_buf, y_buf);
  return success ? { x: x_buf[0], y: y_buf[0] } : null;
}

function set_entity_component_vec2_by_name(
  entity: number,
  component: string,
  x: number,
  y: number
): boolean {
  const id = lib.symbols.flecs_get_component_id_by_name(stringToPtr(component));
  if (!id) return false;
  return lib.symbols.flecs_set_entity_component_vec2(entity, id, x, y);
}

export const flecs = {
  // lifecycle
  init: () => lib.symbols.flecs_init(),
  tick: () => lib.symbols.flecs_progress(),
  destroy: () => lib.symbols.flecs_fini(),

  get_version: () => lib.symbols.flecs_get_version(),

  // entities
  create_entity: (name: string) => lib.symbols.flecs_create_entity(stringToPtr(name)),
  destroy_entity: (entity: number) => lib.symbols.flecs_destroy_entity(entity),

  // component ops
  add_component_by_id: (entity: number, id: number) =>
    lib.symbols.flecs_add_entity_component_by_id(entity, id),
  add_component_by_name: (entity: number, name: string) =>
    lib.symbols.flecs_add_entity_component_by_name(entity, stringToPtr(name)),

  remove_component_by_id: (entity: number, id: number) =>
    lib.symbols.flecs_remove_entity_component_by_id(entity, id),
  remove_component_by_name: (entity: number, name: string) =>
    lib.symbols.flecs_remove_entity_component_by_name(entity, stringToPtr(name)),

  has_component_by_id: (entity: number, id: number) =>
    lib.symbols.flecs_has_entity_component_by_id(entity, id),
  has_component_by_name: (entity: number, name: string) =>
    lib.symbols.flecs_has_entity_component_by_name(entity, stringToPtr(name)),

  get_component_id_by_name: (name: string) =>
    lib.symbols.flecs_get_component_id_by_name(stringToPtr(name)),

  mark_component_changed_by_name: (entity: number, name: string) =>
    lib.symbols.flecs_mark_component_changed_by_name(entity, stringToPtr(name)),
  mark_component_changed_by_id: (entity: number, id: number) =>
    lib.symbols.flecs_mark_component_changed_by_id(entity, id),

  // generic vec2 API
  set_entity_component_vec2_by_name,
  get_entity_component_vec2_by_name,

  // debug helpers
  print_component_registry: () => lib.symbols.flecs_print_component_registry(),
  print_entity_components: (entity: number) =>
    lib.symbols.flecs_print_entity_components(entity),  

  // direct helpers
  set_entity_position: (id: number, x: number, y: number) =>
    lib.symbols.flecs_set_entity_position(id, x, y),
  get_entity_position: (id: number) =>
    get_vec2(id, lib.symbols.flecs_get_entity_position),

  set_entity_velocity: (id: number, x: number, y: number) =>
    lib.symbols.flecs_set_entity_velocity(id, x, y),
  get_entity_velocity: (id: number) =>
    get_vec2(id, lib.symbols.flecs_get_entity_velocity),

  set_entity_destination: (id: number, x: number, y: number, speed: number) =>
    lib.symbols.flecs_set_entity_destination(id, x, y, speed),
  get_entity_destination: (id: number) =>
    lib.symbols.flecs_get_entity_destination(id, x_buf, y_buf, z_buf) ? { x: x_buf[0], y: y_buf[0], speed: z_buf[0] } : null
};

File: src/hxcore/flecs/flecs.nim
# lib/flecs/flecs.nim

# header files are in lib/flecs/flecs_wrapper/include
# source files are in lib/flecs/flecs_wrapper/src

{.passC: "-Ilib/flecs/flecs_wrapper/include".}
{.passC: "-I./include".} 
{.passC: "-std=c99".} 
{.passC: "-D_GNU_SOURCE".}

# use the libflecs_wrapper.a
#[
{.passL: "-static".}
{.passL: "-Llib/flecs/flecs_wrapper/lib".}
{.passL: "-lflecs_wrapper".}
{.passL: "-lm".} # order matters, link math after flecs
]#

# include the source directly
{.compile: "flecs_wrapper/src/flecs.c".}
{.compile: "flecs_wrapper/src/flecs_wrapper.c".}
{.compile: "flecs_wrapper/src/flecs_wrapper_world.c".}
{.compile: "flecs_wrapper/src/flecs_wrapper_entity.c".}
{.compile: "flecs_wrapper/src/flecs_wrapper_component.c".}
{.compile: "flecs_wrapper/src/flecs_wrapper_event.c".}
{.compile: "flecs_wrapper/src/systems/move_system.c".}
{.compile: "flecs_wrapper/src/systems/destination_system.c".}


## Event IDs (order matters since these are mapped in the flecs_wrapper.  Note that indecies start at 1, leaving 0 for  unknown event/error)
#    FLECS_HI_COMPONENT_ID + 40, // EcsOnAdd
#    FLECS_HI_COMPONENT_ID + 41, // EcsOnRemove
#    FLECS_HI_COMPONENT_ID + 42, // EcsOnSet
#    FLECS_HI_COMPONENT_ID + 43, // EcsOnDelete
#    FLECS_HI_COMPONENT_ID + 44, // EcsOnDeleteTarget
#    FLECS_HI_COMPONENT_ID + 45, // EcsOnTableCreate
#    FLECS_HI_COMPONENT_ID + 46, // EcsOnTableDelete
##
const
  ecsUnknownEvent*: uint32 = 0
  ecsOnAdd*: uint32 = 1
  ecsOnRemove*: uint32 = 2
  ecsOnSet*: uint32 = 3
  ecsOnDelete*: uint32 = 4
  ecsOnDeleteTarget*: uint32 = 5
  ecsOnTableCreate*: uint32 = 6
  ecsOnTableDelete*: uint32 = 7

proc flecs_component_get_id_by_name*(name: cstring): uint32 {.importc.}
proc flecs_component_is_mark_changed_by_name*(entity_id: uint32, name: cstring): bool {.importc.}
proc flecs_component_is_marked_changed*(entity_id, component_id: uint32): bool {.importc.}
proc flecs_component_is_tag*(component_id: uint32): bool {.importc.}
proc flecs_component_print_registry*() {.importc.}

proc flecs_entity_print_components*(entity_id: uint32) {.importc.}
proc flecs_entity_has_component*(entity_id, component_id: uint32): bool {.importc.}
proc flecs_entity_has_component_by_name*(entity_id: uint32, name: cstring): bool {.importc.}

proc flecs_entity_add_component*(entity_id, component_id: uint32): bool {.importc.}
proc flecs_entity_add_component_by_name*(entity_id: uint32, name: cstring): bool {.importc.}
proc flecs_entity_remove_component*(entity_id, component_id: uint32): bool {.importc.}
proc flecs_entity_remove_component_by_name*(entity_id: uint32, name: cstring): bool {.importc.}

#[
proc flecs_entity_set_velocity*(entity_id: uint32, x, y: float32): bool {.importc.}
proc flecs_entity_get_velocity*(entity_id: uint32, x, y: ptr float32): bool {.importc.}
proc flecs_entity_set_position*(entity_id: uint32, x, y: float32): bool {.importc.}
proc flecs_entity_get_position*(entity_id: uint32, x, y: ptr float32): bool {.importc.}
proc flecs_entity_set_destination*(entity_id: uint32, x, y, speed: float32): bool {.importc.}
proc flecs_entity_get_destination*(entity_id: uint32, x, y, speed: ptr float32): bool {.importc.}

proc flecs_entity_set_component_vec2*(entity_id, component_id: uint32, x, y: float32): bool {.importc.}
proc flecs_entity_get_component_vec2*(entity_id, component_id: uint32, x, y: ptr float32): bool {.importc.}
]#

proc flecs_entity_create*(name: cstring): uint32 {.importc.}
proc flecs_entity_destroy*(entity_id: uint32): bool {.importc.}

type
  FlecsObserverCallback* = proc(entity_id, component_id, event_id : uint32, component_pointer: pointer, component_size: uint32, callback_id: uint32) {.cdecl.}

proc flecs_register_observer*(
  component_ids: ptr uint32,
  num_components: uint32,
  event_ids: ptr uint32,
  num_events: uint32,
  callback: FlecsObserverCallback,
  callback_id: uint32
): bool {.importc.}

#[
type
  TrampolineSystemCallback* = proc(entity_id: uint32, components: ptr pointer, num_components: uint32, callback_id: uint32) {.cdecl.}
proc flecs_register_system*(
  component_ids: ptr uint32,
  num_components: uint32,
  callback: TrampolineSystemCallback,
  callback_id: uint32
): bool {.importc.}
]#

proc flecs_init*() {.importc.}
proc flecs_progress*(delta_time: float32) {.importc.}
proc flecs_fini*() {.importc.}
proc flecs_version*(): cstring {.importc.}

# Observer callback registry, implemented in Nim to help with ease of use
import std/tables
import std/strformat

type
  ObserverCallback = proc(entity, component, event: uint32, component_pointer: pointer, component_size: uint32) {.gcsafe.}
  #SystemCallback = proc(entity: uint32, components: ptr pointer, num_components: uint32, callback_id: uint32) {.gcsafe.}

# Internal state
var
  nextCallbackId {.global.} = 1'u32  # Reserve 0 for invalid ID
  cbMap = initTable[uint32, ObserverCallback]()

# Register a callback and return its ID
proc register_observer_callback(cb: ObserverCallback, id: uint32 = 0'u32): uint32 =
  if cb.isNil:
    echo "⚠️ Warning: tried to register nil callback"
    return 0

  var realId = id
  if realId == 0:
    realId = nextCallbackId
    inc nextCallbackId

  if cbMap.hasKey(realId):
    echo "⚠️ Warning: callback already registered for id ", realId
    return 0

  cbMap[realId] = cb
  return realId

# Called from C — dispatches to registered Nim callbacks
proc flecs_observer_trampoline(entity, component, event: uint32, component_pointer: pointer, component_size: uint32, callback_id: uint32) {.cdecl.} =
  if cbMap.hasKey(callback_id):
    cbMap[callback_id](entity, component, event, component_pointer, component_size)
  else:
    echo "⚠️ No callback registered for id ", callback_id

# Optional utility to unregister
proc flecs_remove_observer_callback*(id: uint32) =
  if cbMap.hasKey(id):
    cbMap.del(id)
    echo "✅ Callback with id ", id, " unregistered."
  else:
    echo "⚠️ Tried to unregister missing callback id ", id

proc flecs_add_observer_callback*(components:openArray[uint32], events:openArray[uint32], callback: ObserverCallback) : uint32 =
  if callback.isNil:
    echo "⚠️ Warning: tried to register nil callback"
    return

  if components.len == 0 or events.len == 0:
    echo "⚠️ Warning: tried to register observer with no components or events"
    return

  let id = register_observer_callback(callback)
  discard flecs_register_observer(
    components[0].addr,
    uint32 components.len,
    events[0].addr,
    uint32 events.len,
    flecs_observer_trampoline,
    id
  )
  return id
